pkg/trace/obfuscate/credit_cards.go

```diff
@@ -0,0 +1,268 @@
+package obfuscate
+
+import (
+	"strings"
+
+	"github.com/DataDog/datadog-agent/pkg/trace/pb"
+)
+
+// ccObfuscator maintains credit card obfuscation state and processing.
+type ccObfuscator struct {
+	luhn bool
+}
+
+func newCreditCardsObfuscator(useLuhn bool) *ccObfuscator {
+	cco := &ccObfuscator{luhn: useLuhn}
+	pb.SetMetaHook(cco.CheckTag)
+	return cco
+}
+
+// CheckTag checks the tag with the given key and val and returns the final
+// value to be assigned to this tag.
+//
+// For example, in this specific use-case, if the val is detected to be a credit
+// card number, "?" will be returned.
+func (cco *ccObfuscator) CheckTag(k, v string) (newval string) {
+	switch k {
+	case "_sample_rate",
+		"_sampling_priority_v1",
+		"error",
+		"error.msg",
+		"error.type",
+		"error.stack",
+		"env",
+		"graphql.field",
+		"graphql.query",
+		"graphql.type",
+		"graphql.operation.name",
+		"grpc.code",
+		"grpc.method",
+		"grpc.request",
+		"http.status_code",
+		"http.method",
+		"runtime-id",
+		"out.host",
+		"out.port",
+		"sampling.priority",
+		"span.type",
+		"span.name",
+		"service.name",
+		"service",
+		"sql.query",
+		"version":
+		// these tags are known to not be credit card numbers
+		return v
+	}
+	if strings.HasPrefix(k, "_dd") {
+		return v
+	}
+	if isCardNumber(v, cco.luhn) {
+		return "?"
+	}
+	return v
+}
+
+// isSensitve checks if b could be a credit card number by checking the digit count and IIN prefix.
+// If validateLuhn is true, the Luhn checksum is also applied to potential candidates.
+func isCardNumber(b string, validateLuhn bool) (ok bool) {
+	//
+	// Just credit card numbers for now, based on:
+	// • https://baymard.com/checkout-usability/credit-card-patterns
+	// • https://www.regular-expressions.info/creditcard.html
+	//
+	if len(b) == 0 {
+		return false
+	}
+	if b[0] != 0x20 && b[0] != 0x2d && (b[0] < 0x30 || b[0] > 0x39) {
+		// fast path: only valid characters are 0-9, space (" ") and dash("-")
+		return false
+	}
+	i := 0                      // byte index for traversing b
+	prefix := 0                 // holds up to b[:6] digits as a numeric value (for example []byte{"523"} becomes int(523)) for checking prefixes
+	count := 0                  // counts digits encountered
+	foundPrefix := false        // reports whether we've detected a valid prefix
+	recdigit := func(_ byte) {} // callback on each found digit; no-op by default (we only need this for Luhn)
+	if validateLuhn {
+		// we need Luhn checksum validation, so we have to take additional action
+		// and record all digits found
+		var buf strings.Builder
```

Comment:
🏃 I noticed this string builder show up in the profile, see below.

![image](https://user-images.githubusercontent.com/15000/135427276-8fd49cf8-f1b3-4572-bb96-305302b84765.png)

I think you can get away with just using a simple `[]byte` slice as a buffer here, like shown below.

```diff
diff --git a/pkg/trace/obfuscate/credit_cards.go b/pkg/trace/obfuscate/credit_cards.go
index e68c6c60f..0cd5c471e 100644
--- a/pkg/trace/obfuscate/credit_cards.go
+++ b/pkg/trace/obfuscate/credit_cards.go
@@ -85,8 +85,8 @@ func isCardNumber(b string, validateLuhn bool) (ok bool) {
 	if validateLuhn {
 		// we need Luhn checksum validation, so we have to take additional action
 		// and record all digits found
-		var buf strings.Builder
-		recdigit = func(b byte) { buf.WriteByte(b) }
+		buf := make([]byte, 0, len(b))
+		recdigit = func(b byte) { buf = append(buf, b) }
 		defer func() {
 			if !ok {
 				// if isCardNumber returned false, it means that b can not be
@@ -94,7 +94,7 @@ func isCardNumber(b string, validateLuhn bool) (ok bool) {
 				return
 			}
 			// potentially a credit card number, run the Luhn checksum
-			ok = luhnValid(buf.String())
+			ok = luhnValid(buf)
 		}()
 	}
 loop:
@@ -147,7 +147,7 @@ loop:
 // • https://en.wikipedia.org/wiki/Luhn_algorithm
 // • https://dev.to/shiraazm/goluhn-a-simple-library-for-generating-calculating-and-verifying-luhn-numbers-588j
 //
-func luhnValid(str string) bool {
+func luhnValid(str []byte) bool {
 	var (
 		sum int
 		alt bool
```

On my machine this seems to speed things up by ~30%.

```
name                 old time/op  new time/op  delta
IsSensitive/luhn-12   155ns ± 2%   111ns ± 1%  -28.28%  (p=0.008 n=5+5)
```

Reply:
Very nice. Thanks. Applied.